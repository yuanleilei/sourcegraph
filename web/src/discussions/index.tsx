import { isExtensionEnabled } from '../../../shared/src/extensions/extension'
import { SettingsCascadeOrError } from '../../../shared/src/settings/settings'

/**
 * Tells whether or not the code discussions extensions is enabled or not.
 */
export function isDiscussionsEnabled(settingsCascade: SettingsCascadeOrError): boolean {
    return true // for use when sideloading
	/////////////// TODO!(sqs)
    return isExtensionEnabled(settingsCascade.final, 'sourcegraph/code-discussions')
}
