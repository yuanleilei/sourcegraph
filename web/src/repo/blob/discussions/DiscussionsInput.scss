.discussions-input {
    padding: 1rem;

    &__text-box {
        min-height: 7rem;
    }
    &__preview {
        padding: 1rem;
        min-height: 7rem;
    }

    &__title {
        margin-bottom: 1rem;
    }

    &__row {
        text-align: right;
        margin-top: 1rem;
    }

    &__error {
        margin-top: 0.5rem;
        display: flex;
    }

    &__error-icon {
        margin-right: 0.25rem;
    }
}

.theme-light {
    .discussions-input {
        &__button {
            color: $color-light-text-4;
        }
    }
}
