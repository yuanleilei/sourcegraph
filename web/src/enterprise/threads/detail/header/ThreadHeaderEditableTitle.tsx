import React, { useCallback, useState } from 'react'
import { NotificationType } from '../../../../../../shared/src/api/client/services/notifications'
import { ExtensionsControllerProps } from '../../../../../../shared/src/extensions/controller'
import * as GQL from '../../../../../../shared/src/graphql/schema'
import { Form } from '../../../../components/Form'
import { updateThread } from '../../../../discussions/backend'
import { ThreadsAreaContext } from '../../global/ThreadsArea'

interface Props extends Pick<ThreadsAreaContext, 'kindIcon'> {
    thread: Pick<GQL.IDiscussionThread, 'id' | 'title'>
    onThreadUpdate: (thread: GQL.IDiscussionThread) => void
    className?: string
    extensionsController: {
        services: {
            notifications: {
                showMessages: Pick<
                    ExtensionsControllerProps<
                        'services'
                    >['extensionsController']['services']['notifications']['showMessages'],
                    'next'
                >
            }
        }
    }

    /** Provided by tests only. */
    _updateThread?: typeof updateThread
}

/**
 * The title in the thread header, which has an edit mode.
 */
// tslint:disable: jsx-no-lambda
export const ThreadHeaderEditableTitle: React.FunctionComponent<Props> = ({
    thread,
    onThreadUpdate,
    kindIcon: KindIcon,
    className = '',
    extensionsController,
    _updateThread = updateThread,
}) => {
    const [state, setState] = useState<'viewing' | 'editing' | 'loading'>('viewing')
    const [uncommittedTitle, setUncommittedTitle] = useState(thread.title)

    const onSubmit: React.FormEventHandler = async e => {
        e.preventDefault()
        setState('loading')
        try {
            const updatedThread = await _updateThread({ ThreadID: thread.id, title: uncommittedTitle })
            onThreadUpdate(updatedThread)
            setUncommittedTitle(updatedThread.title)
        } catch (err) {
            extensionsController.services.notifications.showMessages.next({
                message: `Error editing thread title: ${err.message}`,
                type: NotificationType.Error,
            })
        } finally {
            setState('viewing')
        }
    }

    const onEditClick = useCallback(() => setState('editing'), [])
    const onCancelClick = useCallback<React.MouseEventHandler>(
        e => {
            e.preventDefault()
            setState('viewing')
            setUncommittedTitle(thread.title)
        },
        [thread.title]
    )

    return state === 'viewing' ? (
        <div className={`d-flex align-items-start justify-content-between flex-wrap ${className}`}>
            <h1 className="font-weight-normal mb-0">
                <KindIcon className="icon-inline text-muted small" /> {thread.title}
            </h1>
            <button type="button" className="btn btn-secondary btn-sm mt-2" onClick={onEditClick}>
                Edit
            </button>
        </div>
    ) : (
        <Form className={`form d-flex ${className}`} onSubmit={onSubmit}>
            <input
                type="text"
                className="form-control form-control-lg flex-1 mr-2"
                value={uncommittedTitle}
                onChange={e => setUncommittedTitle(e.currentTarget.value)}
                placeholder="Title"
                autoComplete="off"
                autoFocus={true}
                disabled={state === 'loading'}
            />
            <div className="text-nowrap flex-0 d-flex align-items-center">
                <button type="submit" className="btn btn-success btn-lg" disabled={state === 'loading'}>
                    Save
                </button>
                <button type="reset" className="btn btn-link" onClick={onCancelClick} disabled={state === 'loading'}>
                    Cancel
                </button>
            </div>
        </Form>
    )
}
