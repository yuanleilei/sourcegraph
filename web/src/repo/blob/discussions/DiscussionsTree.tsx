import * as H from 'history'
import * as React from 'react'
import { Subscription } from 'rxjs'
import { ExtensionsControllerProps } from '../../../../../shared/src/extensions/controller'
import * as GQL from '../../../../../shared/src/graphql/schema'
import { DiscussionsList } from '../../../discussions/DiscussionsList'
import { registerDiscussionsContributions } from './contributions'
import { DiscussionsCreate } from './DiscussionsCreate'
import { DiscussionsThread } from './DiscussionsThread'

interface Props extends ExtensionsControllerProps {
    repoID: GQL.ID
    repoName: string
    commitID: string
    rev: string | undefined
    filePath: string
    history: H.History
    location: H.Location
    compact: boolean
}

export class DiscussionsTree extends React.PureComponent<Props> {
    private subscriptions = new Subscription()

    public componentDidMount(): void {
        this.subscriptions.add(registerDiscussionsContributions(this.props))
    }

    public componentWillUnmount(): void {
        this.subscriptions.unsubscribe()
    }

    public render(): JSX.Element | null {
        const hash = new URLSearchParams(this.props.location.hash.slice('#'.length))
        const threadID = hash.get('threadID') as GQL.ID
        const commentID = hash.get('commentID') as GQL.ID

        if (threadID) {
            return (
                <DiscussionsThread
                    threadID={threadID}
                    commentID={commentID}
                    {...this.props}
                    showNavbar={true}
                    forceURL={true}
                />
            )
        }
        if (threadID) {
            return <DiscussionsThread threadID={threadID} {...this.props} showNavbar={true} forceURL={true} />
        }
        if (hash.get('createThread') === 'true') {
            return <DiscussionsCreate {...this.props} showNavbar={true} />
        }
        return <DiscussionsList {...this.props} />
    }
}
