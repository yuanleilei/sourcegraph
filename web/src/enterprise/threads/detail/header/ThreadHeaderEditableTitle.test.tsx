import React from 'react'
import renderer, { act } from 'react-test-renderer'
import { Subject } from 'rxjs'
import { take } from 'rxjs/operators'
import sinon from 'sinon'
import { NotificationType } from '../../../../../../shared/src/api/client/services/notifications'
import * as GQL from '../../../../../../shared/src/graphql/schema'
import { updateThread } from '../../../../discussions/backend'
import { ThreadHeaderEditableTitle } from './ThreadHeaderEditableTitle'

describe('ThreadHeaderEditableTitle', () => {
    const create = ({
        showMessagesNext = sinon.spy(),
        _updateThread = sinon.spy(),
    }: {
        showMessagesNext?: () => void
        _updateThread?: typeof updateThread
    } = {}) =>
        renderer.create(
            <ThreadHeaderEditableTitle
                thread={{ id: 'a', title: 't' }}
                // tslint:disable-next-line: jsx-no-lambda
                kindIcon={(props: any) => <span {...props} />}
                onThreadUpdate={sinon.spy()}
                className="c"
                extensionsController={{
                    services: {
                        notifications: {
                            showMessages: { next: showMessagesNext },
                        },
                    },
                }}
                _updateThread={_updateThread}
            />
        )

    test('viewing state', () => expect(create().toJSON()).toMatchSnapshot())

    test('enter editing state', () => {
        const e = create()
        act(() => e.root.findByType('button').props.onClick())
        expect(e.toJSON()).toMatchSnapshot()
        expect(e.root.findByType('input')).toBeTruthy()
        expect(e.root.findByType('input').props.value).toBe('t')
    })

    test('edit then immediately cancel', () => {
        const e = create()
        const viewingSnapshot = e.toJSON()

        act(() => e.root.findByType('button').props.onClick())
        expect(e.root.findByType('input')).toBeTruthy()

        act(() => e.root.findByProps({ type: 'reset' }).props.onClick({ preventDefault: () => void 0 }))
        expect(e.toJSON()).toEqual(viewingSnapshot)
        expect(e.root.findAllByType('input').length).toBe(0)
    })

    test('edit, change, then cancel', () => {
        const e = create()
        const viewingSnapshot = e.toJSON()

        act(() => e.root.findByType('button').props.onClick())
        act(() => {
            e.root.findByType('input').props.onChange({ currentTarget: { value: 't2' } })
            e.root.findByProps({ type: 'reset' }).props.onClick({ preventDefault: () => void 0 })
        })
        expect(e.toJSON()).toEqual(viewingSnapshot)
        expect(e.root.findAllByType('input').length).toBe(0)

        act(() => e.root.findByType('button').props.onClick()) // edit again
        expect(e.root.findByType('input').props.value).toBe('t')
    })

    test('save', async () => {
        let resolve: (value: GQL.IDiscussionThread) => void
        const updateThreadResult = new Promise<GQL.IDiscussionThread>(resolve_ => {
            resolve = resolve_
        })
        const e = create({
            _updateThread: input => {
                expect(input.ThreadID).toBe('a')
                expect(input.title).toBe('t2')
                return updateThreadResult
            },
        })

        act(() => e.root.findByType('button').props.onClick())
        act(() => e.root.findByType('input').props.onChange({ currentTarget: { value: 't2' } }))
        act(() => {
            e.root.findByType('form').props.onSubmit({ preventDefault: () => void 0 })
        })

        // Loading.
        expect(e.root.findByType('input').props.disabled).toBeTruthy()
        expect(e.root.findByProps({ type: 'submit' }).props.disabled).toBeTruthy()
        expect(e.root.findByProps({ type: 'reset' }).props.disabled).toBeTruthy()

        // Successfully edited.
        //
        // TODO: This prints harmless warnings like `Warning: An update to null inside a test was
        // not wrapped in act(...).`. The fix is to use async `act` when it is in the released
        // version of React (see https://github.com/facebook/react/issues/14775).
        act(() => {
            resolve({ title: 't2' } as any)
        })
        await updateThreadResult
        expect(e.root.findAllByType('input').length).toBe(0)
    })

    test('error saving', async () => {
        let reject: (error: any) => void
        const updateThreadResult = new Promise<GQL.IDiscussionThread>((_resolve, reject_) => {
            reject = reject_
        })

        const showMessagesNext = sinon.spy()
        const e = create({ _updateThread: () => updateThreadResult, showMessagesNext })

        act(() => e.root.findByType('button').props.onClick())
        act(() => e.root.findByType('input').props.onChange({ currentTarget: { value: 't2' } }))
        act(() => {
            e.root.findByType('form').props.onSubmit({ preventDefault: () => void 0 })
        })

        // TODO: This prints harmless warnings like `Warning: An update to null inside a test was
        // not wrapped in act(...).`. The fix is to use async `act` when it is in the released
        // version of React (see https://github.com/facebook/react/issues/14775).
        act(() => {
            reject({ message: 'x' })
        })
        await updateThreadResult.catch(() => void 0)
        expect(showMessagesNext.callCount).toBe(1)
        expect(showMessagesNext.firstCall.args).toEqual([
            {
                message: 'Error editing title of thread: x',
                type: NotificationType.Error,
            },
        ])
        expect(e.root.findAllByType('input').length).toBe(0)
    })
})
