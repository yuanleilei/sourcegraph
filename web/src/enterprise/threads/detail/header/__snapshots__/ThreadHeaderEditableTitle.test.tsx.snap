// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreadHeaderEditableTitle enter editing state 1`] = `
<form
  className="form d-flex c "
  onInvalid={[Function]}
  onSubmit={[Function]}
>
  <input
    autoComplete="off"
    autoFocus={true}
    className="form-control form-control-lg flex-1 mr-2"
    disabled={false}
    onChange={[Function]}
    placeholder="Title"
    type="text"
    value="t"
  />
  <div
    className="text-nowrap flex-0 d-flex align-items-center"
  >
    <button
      className="btn btn-success btn-lg"
      disabled={false}
      type="submit"
    >
      Save
    </button>
    <button
      className="btn btn-link"
      disabled={false}
      onClick={[Function]}
      type="reset"
    >
      Cancel
    </button>
  </div>
</form>
`;

exports[`ThreadHeaderEditableTitle viewing state 1`] = `
<div
  className="d-flex align-items-start justify-content-between flex-wrap c"
>
  <h1
    className="font-weight-normal"
  >
    <span
      className="icon-inline text-muted small"
    />
     
    t
  </h1>
  <button
    className="btn btn-secondary btn-sm mt-2"
    onClick={[Function]}
    type="button"
  >
    Edit
  </button>
</div>
`;
